server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -758,7 +779,7 @@ private boolean balanceByWeights(NodeSorter sorter) {
          * average. To re-balance we need to move shards back eventually likely
          * to the nodes we relocated them from.
          */
-        private ProjectIndex[] buildWeightOrderedIndices(NodeSorter sorter) {
+        private ProjectIndex[] buildWeightOrderedIndices(NodeSorter sorter) throws NodeSorter.InvalidWeightException {
```

Comment:
Throwing here means the entire balance is skipped if there is a single broken shard. Is that what we want? Or do we want to exclude the broken shard instead?